### DISCUSSION TITLE
RFC - Code generation plugins

### DISCUSSION THREAD

**Discussion by yuval-hazaz** [amplication/amplication#2634] [Ideas]
# RFC - Code generation plugins

## Summary
Amplication generates a very opinionated code, from the tech stack to formatting, naming conventions, and included features the user has little to no control over the result.  

There are many requests from the community for new features and different flavors of the code, that can be implemented as new parameters and settings in Amplication. e.g. new auth providers, support for other DBs, changes to Admin UI, adding support for mono-repo etc. 

Instead of implementing each request separately, and being limited by the number of new features implemented in the system - it is suggested that a new Plugin System be implemented. 

The new plugins system will allow developers from the core team, the community, or third-party applications to add functionality that is decoupled from the platform code and can be easily used to customize the generated code. 

This RFC provides a proposal and details on a method to implement support for the Plugin System.

## High Level
A plugin is custom JavaScript or TypeSciprt code that extends the Amplication Data Service Generator (DSG)with new features.

DSG will include hooks in different lifecycle events that will execute the plugins, providing the context and expecting to get back altered results. 

The altered results can include new files, changes to existing files, and even removal of existing files. 

The users will be able to select which plugins to include for the generation of their app. 

Each app will have a separate list of installed plugins. 

## The plugin
A plugin is an NPM package with custom JavaScript or TypeSciprt code. 
The code should follow specific rules to handle specific events. 
Each plugin will get the context from DSG including the files that were generated to that point, and any other settings known to DSG. 
The plugin will be able to generate new files, alter existing files, and even remove files that were already generated. The plugin will also be able to alter the given settings to impact other plugins that will be triggered later in the process.

## Examples
Following are some examples of functionality that can be achieved by plugins:
 - Add a new authentication provider
 - Replace PostgreSQL with other types of DB
 - Add new schema (tables/ fields) to the DB
 - Add new endpoints to controllers or queries and mutations to resolvers
 - Change the naming convention of generated files
 - Change the folder structure of the generated code
 - Add support for mono-repo like Lerna or nx
 
## Plugins library
Plugins can be created by anyone, from Official plugins created by Amplication, to plugins created by the community, or third parties. 
All plugins need to be registered in the plugin library in order to be available for installation in an app. 

## Running Plugins
Since plugins can be created by anyone, and since plugins execute code - they can be used as an attack vector. 
It is not expected that Amplication will be able to review and approve every plugin. 
So, plugins must be executed in an isolated environment without access to any private resources. For example, they can be executed on services like AWS Code build, or GCP Cloud Build.
Also, it should be considered to totally prevent plugins to use specific APIs like Network (TBC).

**Comment by yuval-hazaz** [comment]
Please share any thoughts about any of the below:
- The technology we can use to implement the plugins architecture
- Security Concerns
- Examples of use cases
- New ideas or features that can be added
- anything you think that may be helpful

**Comment by MichaelSolati** [comment]
Would this allow for plugins to have configuration files? For example. If I was to use a custom set of rules with ESLint, can I create/use a "linting" plugin that ran with the `--fix` flag which I would provide it with my configuration somehow?

Not saying this is 100% necessary, but maybe hosting config files for plugins would be nice.

**Reply by yuval-hazaz** [reply]
> Would this allow for plugins to have configuration files? For example. If I was to use a custom set of rules with ESLint, can I create/use a "linting" plugin that ran with the `--fix` flag which I would provide it with my configuration somehow?
> 
> Not saying this is 100% necessary, but maybe hosting config files for plugins would be nice.

Plugins will be able to expose a set of parameters to be configured by the users when installing the plugin, and they will be available for the plugin in the code generation process. 
The plugin will be able to add any kind of file to the generated app, including configuration files that will be used in runtime

**Reply by MichaelSolati** [reply]
So i could theoretically pass in env variables, but no JSON. That's good to know.

**Reply by yuval-hazaz** [reply]
That is an interesting thought, if we can think of enough use cases that this may be needed we can probably include something like that. It will probably not be in the first release. If you think of any other use case where a JSON will be needed, please share

**Comment by yuval-hazaz** [comment]
related to https://github.com/amplication/amplication/issues/2369

**Comment by kwinyyyc** [comment]
Wonder what would the context exposed by DSG be like.
The plugin system is composable, as a plugin developer, I would need to consider interfacing with the previous plugin, e.g. not overwriting things added, and interfacing with the next plugin, e.g notifying it of the updated context somehow. Not sure how it works.

A bit related to the above, but the sequence of the plugins is also important. As a Amplication user,  I am concerned if I have the knowledge to arrange the sequence of the plugins.

Just thinking out loud.

**Reply by yuval-hazaz** [reply]
@kwinyyyc great points!

the order of plugins is important and will be configurable by the user. I agree that the user may need to have some knowledge to understand the correct order.


The context will include a list of all files (called Modules in DSG) that were already created. The plugin writer will be able to add new modules and update existing ones. 

In addition, the context object will include all the settings and parameters received from Amplication.

We are still planning what else should be there and would love to present the current progress and get your feedback.

let me know if you are interested and we can schedule something for the first week of June. 
By then, I think @EugeneTseitlin would be able to share a link to the branch with some progress

**Reply by kwinyyyc** [reply]
Perhaps plugin developer can provide some sort of input and output "contract" via interface or JSON schema, so Amplication can warn user in case the order is not going to work.

Sure, would be interested to take a look!

**Comment by yuval-hazaz** [comment]
great
you should already have the link to my calendar, feel free to select a time for later this coming week or early next week and I will ask @EugeneTseitlin to also join and present the POC

**Reply by mylesgodino** [reply]
Uu
